_posts/2016-02-07-coupling-in-distributed-systems.md

```diff
@@ -0,0 +1,88 @@
+
+## Runtime approaches
+
+Now that I've been honest with you about my compile bias I can explain different approaches and trade-offs for the schema validation problem.
+
+### Implicit schemas
+
+First runtime approach is the loosest one: using implicit schemas and trusting in the good will of producers. As nobody is checking the validity of messages before of being published into Kafka that means that consumers could blow up.
+
+First corrective measure is assuring that only the processing of the poisoned message will blow and not the whole consumer. An example of that would be providing a resume supervision strategy on [Akka Streams](http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0.2/scala.html) when the message don't hold expected implicit schema.
+
+Second corrective measure would be not simply swallowing those crashes but being able to communicate them to proper actors (being humans or software). Good practice is to provide dead letter queues for poisoned messages just in case we want to manipulate and retry the processing of those messages at that level.
+
+Before of getting into explicit schemas I would say that those measure are not usually enough but they are a good safety net, as shit happens, and we need to be prepared.
+
+### Explicit schemas
+
+If we want to avoid poisoned messages to get into our topics we could provide a middle-man service to intercept and validate explicit schemas. That's an integration service that could be a single point of failure but, at the same time, it could valuable to have a centralised repo of schemas when we have a lot of consumers and the complexity of the system would be hard to grasp in de-centralised fashion.
+
+[Schema registry](http://docs.confluent.io/1.0/schema-registry/docs/index.html) is an example of that for Kafka, and its documentation is full of insights about how to implement that in a distributed, highly available and scalable way.
+
+## Compile approaches
+
+Last approach would be creating software that makes impossible to create messages that not hold the expected schema. Assuming [Scala](http://www.scala-lang.org/), we could create a jar that contains [case classes](http://docs.scala-lang.org/tutorials/tour/case-classes.html) that are object materialisations of a schema.
+
+What are the benefits of this approach?
+
+1. Fail early. We don't have to wait until testing or production to verify that the messages published by some producer are correct.
+2. Centralised knowledge.
+
+What is the solvable problem?
+
+- Cascade updates. If our microservices live in different repos, then we need to make sure that updates into that common binary are applied into producer and consumers. That's cumbersome and if it's not done could generate non expected bugs as we introduced a false sense of security with that library. That could be solved using a [monorepo](http://danluu.com/monorepo/) 
```

Comment:
I think this problem already exists without the schema library. If you update the schema of your REST API and break the previous contract with your client, then the API won't work, no matter wether this contract is enforced through a lib or loosely by implicit schemas and trust.